### DISCUSSION TITLE
Discussion: additional entropy source for Linux kernel

### DISCUSSION THREAD

**Discussion by Snawoot** [drand/drand#926] [General]
Hello,

I think it'll be useful to provide randomness from League of Entropy servers as an entropy source for wide use. Current drand implementation allows to fetch private entropy from group of drand servers, but it has drawbacks:

* There is no straightforward way to put extracted data to good use:
  * Output is not machine readable because JSON is prepended with program messages. Probably it can be easily resolved if daemon would output its messages into stderr and JSON into stdout. (#152)
  * It's up to operator how to properly utilize extracted data to improve his RNG security.
  * It'll require some external runner/wrapper to fetch randomness continuously, restarting program for each request.
* Current implementation of private randomness fetch doesn't utilize distributed nature of the Distributed Randomness Beacon. `drand` [takes first successful response and returns it](https://github.com/dedis/drand/blob/0509cb73af6d4a7bc366d53687351a480f04adda/public.go#L25-L30), relying solely on that source.

Taking all above into account, I've built distributed randomness beacon client, compatible with current `drand` servers: [package](https://pypi.org/project/drb-client/), [source](https://github.com/Snawoot/drb-client). Basically it gathers private entropy from servers of the group and mixes sufficient amount of responses from distinct servers with stateful HKDF mixer and then outputs result into Linux kernel entropy pool. This way every output does not relies on trust to single `drand` server and I think we can provide such guarantees:

* Client set to quorum of `Q` servers out of group of `N` servers guaranteed to produce true randomness if at least `N - Q + 1` servers provide true random numbers.
* Since new HKDF salt derived among with each successful output and kept in secret, output of `drb-client` will become unpredictable after first successful entropy derivation, even if all servers get compromised since that moment.

What do you think about this approach? Is it feasible?

**Comment by nikkolasg** [comment]
Nice project !! Your approach makes total sense: All your points are spot on and were set as next items todo for drand -> drand is meant to incorporate your `drb-clients` features in the future. Reason why these are not present at the moment is mostly due to a lack of free cycles and higher priority issues.

Just one question: would it be easy to port drb-client on MacOS ? does it work already ?

**Comment by Snawoot** [comment]
Thanks!

`stdout` entropy sink should work on all platforms, including Windows. `rndaddentropy` is Linux-specific. `devrandom` should work on all unix-like systems where feeding CSPRNG via `/dev/random` write is possible. I think it is the case for Mac OS also, but I didn't managed to find corresponding source code or documentation on that subject.

I believe code can be extended with any platform-specific sink with little efforts. It is possible to invoke any native function with pure Python code using `ctypes` module. This way no platform-specific builds and packaging are needed, all can be handled with a dynamic nature of language.

I already tested `drb-client` on Windows with success and soon I'll try it on Mac OS. Also I'll try to lookup Darwin source to confirm writes to `/dev/random` mixed into entropy pool as expected (despite not increasing entropy count, it's not strictly necessary). I hope I'll post update with my findings soon.

**Comment by Snawoot** [comment]
Here is update about Mac OS. `drb-client` works on Mac OS, but `devrandom` sink appears to be useless.

Looking at darwin-xnu code I see write into `/dev/random` is legitimate and supposed to mix entropy into pool: https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/dev/random/randomdev.c#L159-L187 And default char dev permissions allow anyone to do that.

But deeper look reveals it is basically a no-op:
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/prng/prng_random.c#L567-L582

And `PERMIT_WRITE_RANDOM` is defined as `0`:
https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/prng/prng_random.c#L249-L261

Only way to add entropy into kernel pool appears to be `prng_addentropy` which most likely is exposed to CoreCrypto. I don't know if it is reachable from user code somehow - I'm not familiar with Mac OS system architecture.

**Comment by nikkolasg** [comment]
Bumping up this issue to make sure we advertise this project on the readme and more if necessary :)

**Comment by Snawoot** [comment]
Hello!

Current version of `drb-client` seem to be out of sync with exchange format of `drand`. It looks like public key format in `group.toml` has been changed and now public key is 48 bytes string instead of 128 bytes like it was before.

What have been changed? Is there any other changes in cryptography for `/api/private` method? Can we expect some stable API without breaking changes?

**Comment by nikkolasg** [comment]
The curve have changed to BLS12-381 now, using the v7 of the hash-to-curve RFC https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07
I'll write a spec of this current version and you should expect it to be stable for a while (unless a major breakthrough in the curve has been found and it's not secure anymore).

**Comment by nikkolasg** [comment]
Actually now that you mention it, it's unclear if the `/api/private` method is gonna be available on all nodes, given it's a potential for a DDOS: it forces the node to to asymmetrical encryption which is expensive without any rate limit. 
Will try to keep this issue updated with any news on that front.

**Comment by nikkolasg** [comment]
Also, I'm thinking that this method potentially don't need to use the same curve as the public randomness part. It could just be ed25519 FWIW, and this would be completely stable.
A client sends its ephemereal public key, the node answers back by encrypting to this eph. public key. That's it, no need to have the same curve as the public randomness at all.

**Comment by Snawoot** [comment]
Curve25519 appears much more preferable to me because it's very challenging to find proven and fast implementation of rare curves (such as BN256 or BLS12-381) for various programming languages/runtimes and environments.

External RNG makes much sense for tiny devices w/o own reliable HW RNG, and some lightweight well-known curve like 25519 with well-optimized implementation shall fit well. Currently BN256 pairing in `drb-client` takes about 1 second of CPU time per request. Yes, it's slow Python implementation of BN256, but there was not much of a choice.

> Actually now that you mention it, it's unclear if the /api/private method is gonna be available on all nodes, given it's a potential for a DDOS: it forces the node to to asymmetrical encryption which is expensive without any rate limit.

I think it's OK if some nodes will fail to answer: it's intended by design to expect at least (`quorum`) different inputs from distinct nodes.
